### PR TITLE
Add `contains_over_range_not_nil` rule

### LINKED ISSUE #2776: Rule request: use contains() instead of range(of:) != nil
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

Similar to `contains_over_first_not_nil` (https://github.com/realm/SwiftLint/pull/1846) but for strings

1. **Why should this rule be added?** It’s more idiomatic to ask whether the collection contains the element. Checking whether the range of the substring is nil is a more Obj-C way of doing it (e.g. with `range.location == NSNotFound`)
2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

Triggering:

```swift
if items.range(of: item) != nil {
    // items contains item
}

if items.range(of: item) == nil { // notice the negation of the previous example
    // items does not contain item
}
```

Not triggering:

```swift
if item.contains(item) {
    // items contains item
}

if !items.contains(item) { // notice the negation of the previous example
    // items does not contain item
}
```
3. **Should the rule be configurable, if so what parameters should be configurable?** I don’t think it needs to be configurable.
4. **Should the rule be opt-in or enabled by default? Why?** I think it should be on by default. It’s always clearer, in my opinion, to express your intent in checking containment, rather than grabbing the index and checking if it’s nil. Plus, it saves on the calculation and storage allocation for the range, which isn’t much, but is probably more than a `Bool`.

### PR DESCRIPTION
**Pull request by cltnschlosser**
Also make `contains_over_first_not_nil` also match == nil

Resolves #2776

### CODE REVIEW

File: Tests/LinuxMain.swift [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Generated using Sourcery 0.16.0 — https://github.com/krzysztofzablocki/Sourcery
+// Generated using Sourcery 0.16.1 — https://github.com/krzysztofzablocki/Sourcery
```
Comment (cltnschlosser) [inline Tests/LinuxMain.swift]: I ran sourcery, but tests weren't generated, suggestions?

**Comment by marcelofabri** [comment]
This should only match `contains` with a single param (https://github.com/mozilla-mobile/firefox-ios/blob/4d52a5297ee08ee69f0c366e0aa8226093bfa086/content-blocker-lib-ios/src/TrackingProtectionPageStats.swift#L192:16 is a false positive)

File: Tests/LinuxMain.swift (hunk 1 above)
Comment (cltnschlosser) [reply Tests/LinuxMain.swift]: I was missing `AutomaticTestableRule`

**Comment by cltnschlosser** [comment]
@marcelofabri Added a check for argument count

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,7 +10,12 @@
 
 #### Enhancements
 
-* None.
+* Add `contains_over_range_not_nil` rule to prefer using `contains` over comparison of `range(of:)` to `nil`.
```
Comment (marcelofabri) [inline CHANGELOG.md]: same as the other PR, can you please mention that this is an opt-in rule and add `#` before the issue number?

File: Source/SwiftLintFramework/Rules/Performance/ContainsOverRangeNotNilRule.swift [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            ]
+        }
+    )
+
+    public func validate(file: File) -> [StyleViolation] {
+        let pattern = "\\)\\s*(==|!=)\\s*nil"
+        return validate(file: file, pattern: pattern, patternSyntaxKinds: [.keyword],
+                        callNameSuffix: ".range", severity: configuration.severity,
+                        reason: "Prefer `contains` over `range(of:) != nil`") { expression in
+                            guard let substructure = expression["key.substructure"]
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Performance/ContainsOverRangeNotNilRule.swift]: you should the accessors defined in `Dictionary+SwiftLint` instead:

```
return expression.enclosedArguments.count == 1
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -23,6 +23,13 @@
   expressions like `filter(where:).isEmpty` instead of `contains(where:)`.  
   [Marcelo Fabri](https://github.com/marcelofabri)
 
+* Add `contains_over_range_not_nil` opt-in rule to prefer using `contains` over comparison of `range(of:)` to `nil`.  
+  [Colton Schlosser](https://github.com/cltnschlosser)
+  [#2776](https://github.com/realm/SwiftLint/issues/2776)
+
+* Make `contains_over_first_not_nil` rule also match `first(where:) == nil`.
```
Comment (marcelofabri) [inline CHANGELOG.md]: Maybe we should split this into a different PR and discuss there? I'm a bit conflicted because while I agree that matching `== nil` is useful, the rule is named `contains_over_first_not_nil`.

File: CHANGELOG.md [hunk 5]
```diff
@@ -23,6 +23,13 @@
   expressions like `filter(where:).isEmpty` instead of `contains(where:)`.  
   [Marcelo Fabri](https://github.com/marcelofabri)
 
+* Add `contains_over_range_not_nil` opt-in rule to prefer using `contains` over comparison of `range(of:)` to `nil`.
```
Comment (marcelofabri) [inline CHANGELOG.md]: Can we rename the rule to `contains_over_range_nil_comparison` or something like that?

File: Source/SwiftLintFramework/Rules/Performance/ContainsOverRangeNotNilRule.swift [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+public struct ContainsOverRangeNotNilRule: CallPairRule, OptInRule, ConfigurationProviderRule, AutomaticTestableRule {
+    public var configuration = SeverityConfiguration(.warning)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "contains_over_range_not_nil",
+        name: "Contains over range not nil",
+        description: "Prefer `contains` over `range(of:) != nil`.",
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Performance/ContainsOverRangeNotNilRule.swift]: if we rename the rule, we should also update the description

File: Source/SwiftLintFramework/Rules/Performance/ContainsOverRangeNotNilRule.swift [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+            ]
+        }
+    )
+
+    public func validate(file: File) -> [StyleViolation] {
+        let pattern = "\\)\\s*(==|!=)\\s*nil"
+        return validate(file: file, pattern: pattern, patternSyntaxKinds: [.keyword],
+                        callNameSuffix: ".range", severity: configuration.severity,
+                        reason: "Prefer `contains` over `range(of:) != nil`") { expression in
+                            return expression.enclosedArguments.count == 1
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Performance/ContainsOverRangeNotNilRule.swift]: on a second thought, can we compare the param name too to be extra careful?

This should do the trick: 

```swift
return expression.enclosedArguments.map { $0.name } == ["of"]
```

Comment (cltnschlosser) [reply Source/SwiftLintFramework/Rules/Performance/ContainsOverRangeNotNilRule.swift]: I'll update with the rename and the parameter check improvement.

Comment (cltnschlosser) [reply Source/SwiftLintFramework/Rules/Performance/ContainsOverRangeNotNilRule.swift]: Done

**Comment by jpsim** [comment]
`String.contains(_: String)` only works if you import Foundation.

```
$ echo "import Foundation\nprint(\"asdf\".contains(\"as\"))" > test.swift
$ swift test.swift 
true
$ echo "print(\"asdf\".contains(\"as\"))" > test.swift 
$ swift test.swift                                    
test.swift:1:23: error: cannot convert value of type 'String' to expected argument type 'String.Element' (aka 'Character')
print("asdf".contains("as"))
                      ^~~~
```

**Comment by cltnschlosser** [comment]
@jpsim `range(of:)` requires Foundation as well.
